### PR TITLE
Use DB backend in test client

### LINKED ISSUE #1603: test_client should use `client-db` backed by an in-memory KVDB
Currently the `test_client` is using the `in_mem` implementation of `client::Backend`. The actual `client-db` should be used, but with a `kvdb-memorydb`.

https://github.com/paritytech/substrate/blob/1a26bd70b7d7b1f6411558c81753d607883566ba/core/test-client/src/lib.rs#L65

### PR DESCRIPTION
**Pull request by svyatonik**
closes #1603 

Also fixed two issues, revealed by this migration:
1) wrong assumptions about API safety in changes trie storage implementation;
2) fixed simultaneous import of block B + finalization of its ancestors (i.e. #1602).

This PR also introduced `test_client::new_light` to be used in integration tests. The `test_client::LightFetcher` is currently a stub, because I'm not sure yet about its implementation (it will be fixed as a part of #1669).

### CODE REVIEW

File: core/client/db/src/lib.rs [hunk 1]
```diff
@@ -463,8 +466,10 @@ impl<Block: BlockT> client::backend::PrunableStateChangesTrieStorage<Blake2Hashe
 
 impl<Block: BlockT> state_machine::ChangesTrieRootsStorage<Blake2Hasher> for DbChangesTrieStorage<Block> {
 	fn root(&self, anchor: &state_machine::ChangesTrieAnchorBlockId<H256>, block: u64) -> Result<Option<H256>, String> {
-		// check API requirement
-		assert!(block <= anchor.number, "API requirement");
+		// check API requirement: we can't get NEXT block(s) based on anchor
+		if block > anchor.number {
+			return Err(format!("Can't get CT root at {} using anchor at {}", block, anchor.number));
```
Comment (pepyakin) [inline core/client/db/src/lib.rs]:
```suggestion
			return Err(format!("Can't get changes trie root at {} using anchor at {}", block, anchor.number));
```

this, I believe, should be slightly more clear for users.

**Comment by bkchr** [comment]
Can we not then drop `client/in_mem.rs`?

**Comment by bkchr** [comment]
But the kvdb in memory implementation is also just in memory? In my opinion this just requires additional code that is only used in tests and we don't test the real code that interacts with the db interface.
We should maybe improve the implementation of the kvdb in memory to support stuff like comparing etc.

**Comment by svyatonik** [comment]
> But the kvdb in memory implementation is also just in memory?

Comparing kvdb = comparing db-level representation (columns, ...) even if we improve it. Also it could be affected by some irrelevant factors (insertion order - idk?). To compare two client backends we need to compare the set of headers + the set of chain (best + last finalized + ...) + ... => we'll need several `HashMap`s + insertion ops anyway (which is `in_mem::*`).

> In my opinion this just requires additional code that is only used in tests and we don't test the real code that interacts with the db interface.

Exactly. We have a lot of code that is only used in tests (helpers/stubs/mock objects/...). Consider `in_mem::*` as another one helper. Like we do not tend to use wasm executor in **every** test, though there are some tests that are testing it.

**Comment by svyatonik** [comment]
That said, I'm ready to give up if you'll find more supporters on your side :) But would prefer to separate "Use DB backend in test-client" and "Get rid of in-memory backend" tasks.

**Comment by bkchr** [comment]
I don't want you to give up^^ 
It is just that `kvdb-memorydb` uses a hashmap internally: https://github.com/paritytech/parity-common/blob/9abaf267d1319ed65e2fdd90ee55432dc4540f78/kvdb-memorydb/src/lib.rs#L29

In #1609 we have the situation that we need to implement a completely different interface for `in_mem`. Thus, the real interface will not be tested by unit tests. So, we maybe do not catch bugs in tests.

**Comment by svyatonik** [comment]
I'd say that it is probably better to rebase #1609 on this PR && add `unimplemented!()` in `in_mem::Blockchain::children()` until some specific implementation is actually required for some low-level tests. I could see that right now all new tests there are using `test_client::new()` => DB backend will be used.

**Comment by bkchr** [comment]
Okay, sounds like a plan :)

File: core/client/db/src/lib.rs [hunk 2]
```diff
@@ -531,8 +536,8 @@ impl<Block: BlockT<Hash=H256>> Backend<Block> {
 		Backend::from_kvdb(db as Arc<_>, config.pruning, canonicalization_delay)
 	}
 
-	#[cfg(test)]
-	fn new_test(keep_blocks: u32, canonicalization_delay: u64) -> Self {
+	#[cfg(any(test, feature = "test-helpers"))]
+	pub fn new_test(keep_blocks: u32, canonicalization_delay: u64) -> Self {
```
Comment (bkchr) [inline core/client/db/src/lib.rs]: As described in the issue, this should use `kvdb-inmemory`.

Comment (svyatonik) [reply core/client/db/src/lib.rs]: It uses it, no? A couple of lines below:
```
let db = Arc::new(::kvdb_memorydb::create(NUM_COLUMNS));
```

Comment (bkchr) [reply core/client/db/src/lib.rs]: Ohh :( Sorry!

**Pull request merged**